File: backend/runtime/internal/config/crypto.go
package config

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"io"

	"github.com/zalando/go-keyring"
)

const (
	serviceName = "LunaAI"
	keyName     = "MasterKey"
)

// CryptoService 提供基于 OS Keychain 的 AES-256-GCM 加解密服务
type CryptoService struct {
	masterKey []byte
}

// NewCryptoService 初始化 CryptoService，从 Keychain 获取或生成 Master Key
func NewCryptoService() (*CryptoService, error) {
	keyStr, err := keyring.Get(serviceName, keyName)
	var masterKey []byte

	if err != nil {
		// 如果找不到，生成一个新的 32 字节 (256-bit) 密钥
		masterKey = make([]byte, 32)
		if _, err := io.ReadFull(rand.Reader, masterKey); err != nil {
			return nil, fmt.Errorf("生成 Master Key 失败: %w", err)
		}

		// 将生成的密钥进行 Base64 编码后存入 Keychain
		keyStr = base64.StdEncoding.EncodeToString(masterKey)
		if err := keyring.Set(serviceName, keyName, keyStr); err != nil {
			return nil, fmt.Errorf("保存 Master Key 到 Keychain 失败: %w", err)
		}
	} else {
		// 从 Keychain 读取并解码
		masterKey, err = base64.StdEncoding.DecodeString(keyStr)
		if err != nil {
			return nil, fmt.Errorf("解码 Master Key 失败: %w", err)
		}
		if len(masterKey) != 32 {
			return nil, fmt.Errorf("Master Key 长度不正确，期望 32 字节，实际 %d 字节", len(masterKey))
		}
	}

	return &CryptoService{
		masterKey: masterKey,
	}, nil
}

// Encrypt 使用 Master Key 对明文进行 AES-256-GCM 加密
func (s *CryptoService) Encrypt(plaintext string) (string, error) {
	block, err := aes.NewCipher(s.masterKey)
	if err != nil {
		return "", fmt.Errorf("创建 AES cipher 失败: %w", err)
	}

	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("创建 GCM 失败: %w", err)
	}

	nonce := make([]byte, aesGCM.NonceSize())
	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
		return "", fmt.Errorf("生成 nonce 失败: %w", err)
	}

	ciphertext := aesGCM.Seal(nonce, nonce, []byte(plaintext), nil)
	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

// Decrypt 使用 Master Key 对密文进行 AES-256-GCM 解密
func (s *CryptoService) Decrypt(ciphertextBase64 string) (string, error) {
	ciphertext, err := base64.StdEncoding.DecodeString(ciphertextBase64)
	if err != nil {
		return "", fmt.Errorf("解码密文失败: %w", err)
	}

	block, err := aes.NewCipher(s.masterKey)
	if err != nil {
		return "", fmt.Errorf("创建 AES cipher 失败: %w", err)
	}

	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("创建 GCM 失败: %w", err)
	}

	nonceSize := aesGCM.NonceSize()
	if len(ciphertext) < nonceSize {
		return "", fmt.Errorf("密文长度太短")
	}

	nonce, ciphertext := ciphertext[:nonceSize], ciphertext[nonceSize:]
	plaintext, err := aesGCM.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return "", fmt.Errorf("解密失败: %w", err)
	}

	return string(plaintext), nil
}

File: backend/ai-service/app/config/crypto.py
"""
Luna AI 加解密服务模块

做什么：提供基于 OS Keychain 的 AES-256-GCM 加解密服务。
为什么这样做：保护 API Key 等敏感信息，避免明文落盘。
输入输出：
    - CryptoService: 加解密服务类
边界条件：
    - 首次运行时生成 32 字节 Master Key 并存入 Keychain
    - 后续运行从 Keychain 读取 Master Key
异常行为：
    - Keychain 访问失败或加解密失败时抛出异常
"""

import base64
import os

import keyring
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from app.logger import logger

SERVICE_NAME = "LunaAI"
KEY_NAME = "MasterKey"


class CryptoService:
    """提供基于 OS Keychain 的 AES-256-GCM 加解密服务"""

    def __init__(self):
        self.master_key = self._get_or_create_master_key()

    def _get_or_create_master_key(self) -> bytes:
        """从 Keychain 获取或生成 Master Key"""
        try:
            key_str = keyring.get_password(SERVICE_NAME, KEY_NAME)
            if key_str:
                master_key = base64.b64decode(key_str)
                if len(master_key) != 32:
                    raise ValueError(f"Master Key 长度不正确，期望 32 字节，实际 {len(master_key)} 字节")
                return master_key
        except Exception as e:
            logger.warning(f"从 Keychain 读取 Master Key 失败或不存在，将生成新密钥: {e}")

        # 如果找不到或读取失败，生成一个新的 32 字节 (256-bit) 密钥
        master_key = os.urandom(32)
        key_str = base64.b64encode(master_key).decode('utf-8')
        
        try:
            keyring.set_password(SERVICE_NAME, KEY_NAME, key_str)
            logger.info("已生成新的 Master Key 并保存到 Keychain")
        except Exception as e:
            logger.error(f"保存 Master Key 到 Keychain 失败: {e}")
            # 降级：如果无法保存到 Keychain，仅在内存中保留（重启后失效）
            # 实际生产环境中可能需要更严格的处理
            
        return master_key

    def encrypt(self, plaintext: str) -> str:
        """使用 Master Key 对明文进行 AES-256-GCM 加密"""
        if not plaintext:
            return ""
            
        aesgcm = AESGCM(self.master_key)
        nonce = os.urandom(12) # GCM 推荐的 nonce 长度为 12 字节
        
        ciphertext = aesgcm.encrypt(nonce, plaintext.encode('utf-8'), None)
        
        # 将 nonce 和 ciphertext 拼接后进行 Base64 编码
        # Go 版本的 Seal 会将 ciphertext 附加到 nonce 后面
        combined = nonce + ciphertext
        return base64.b64encode(combined).decode('utf-8')

    def decrypt(self, ciphertext_base64: str) -> str:
        """使用 Master Key 对密文进行 AES-256-GCM 解密"""
        if not ciphertext_base64:
            return ""
            
        try:
            combined = base64.b64decode(ciphertext_base64)
            
            if len(combined) < 12:
                raise ValueError("密文长度太短")
                
            nonce = combined[:12]
            ciphertext = combined[12:]
            
            aesgcm = AESGCM(self.master_key)
            plaintext = aesgcm.decrypt(nonce, ciphertext, None)
            
            return plaintext.decode('utf-8')
        except Exception as e:
            raise ValueError(f"解密失败: {e}")
